### PR TITLE
Implement ColorDegeneration KPL

### PR DESCRIPTION
**Pull request by LukeWood**
color_degeneration

### CODE REVIEW

**Comment by qlzh727** [comment]
https://github.com/keras-team/keras-cv/issues/86

File: examples/layers/preprocessing/random_color_degeneration_demo.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
+"""random_shear_demo.py shows how to use the RandomShear preprocessing layer.
```
Comment (qlzh727) [inline examples/layers/preprocessing/random_color_degeneration_demo.py:1]: do we need a license header for this?

File: keras_cv/layers/preprocessing/__init__.py [hunk 2]
```diff
@@ -39,3 +39,4 @@
 from keras_cv.layers.preprocessing.random_sharpness import RandomSharpness
 from keras_cv.layers.preprocessing.random_shear import RandomShear
 from keras_cv.layers.preprocessing.solarization import Solarization
+from keras_cv.layers.preprocessing.random_color_degeneration import RandomColorDegeneration
```
Comment (qlzh727) [inline keras_cv/layers/preprocessing/__init__.py]: This should be sorted.

File: keras_cv/layers/preprocessing/random_color_degeneration_test.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import tensorflow as tf
+
+from keras_cv.layers import preprocessing
+
+
+class RandomColorDegenerationTest(tf.test.TestCase):
+    def test_random_color_degeneration_base_case(self):
```
Comment (qlzh727) [inline keras_cv/layers/preprocessing/random_color_degeneration_test.py:20]: we should add more solid test for the actual layer behavior, eg validating the output value.

File: keras_cv/utils/preprocessing.py [hunk 4]
```diff
@@ -91,6 +91,23 @@ def blend(image1: tf.Tensor, image2: tf.Tensor, factor: float) -> tf.Tensor:
     temp = image1 + scaled
     return tf.clip_by_value(temp, 0.0, 255.0)
 
+def parse_value_range(param, min_value=0.0, max_value=1.0, param_name="factor"):
```
Comment (qlzh727) [inline keras_cv/utils/preprocessing.py:95]: Probably should call it parse_factor_value_range (disambiguous about the input value range)

File: keras_cv/layers/preprocessing/random_color_degeneration.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+    def get_random_transformation(self):
+        if self.factor[0] == self.factor[1]:
+            return self.factor[0]
+        return self._random_generator.random_uniform(
+            (), self.factor[0], self.factor[1], dtype=tf.float32
+        )
+
+    def augment_image(self, image, transformation=None):
+        degenerate = tf.image.grayscale_to_rgb(tf.image.rgb_to_grayscale(image))
```
Comment (qlzh727) [inline keras_cv/layers/preprocessing/random_color_degeneration.py:56]: Does the tf.image op has any assumption about the input value range?

**Review by qlzh727 — CHANGES_REQUESTED** [review]

File: examples/layers/preprocessing/random_color_degeneration_demo.py (hunk 1 above)
Comment (LukeWood) [reply examples/layers/preprocessing/random_color_degeneration_demo.py:14]: Updated and updated script to include this.

File: keras_cv/layers/preprocessing/random_color_degeneration.py (hunk 5 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/random_color_degeneration.py:56]: Nope:
https://github.com/tensorflow/tensorflow/blob/v2.8.0/tensorflow/python/ops/image_ops_impl.py#L2515

It does a standard LUMA conversion which is just a weighted product.

Comment (LukeWood) [reply keras_cv/layers/preprocessing/random_color_degeneration.py:56]: (R*0.59 + G*someconstant + b*someconstant)

File: keras_cv/utils/preprocessing.py (hunk 4 above)
Comment (LukeWood) [reply keras_cv/utils/preprocessing.py]: Done

File: keras_cv/layers/preprocessing/random_color_degeneration_test.py (hunk 3 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/random_color_degeneration_test.py:20]: Good point - pretty easy to compute some values using luma anyways.  done.

File: keras_cv/layers/preprocessing/__init__.py (hunk 2 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/__init__.py]: Done

**Comment by LukeWood** [comment]
ok - fixed linting failures.  @qlzh727

**Review by qlzh727 — APPROVED** [review]

**Pull request merged**